File: test_projects/ts_pipeline/src/processor.ts
// SPDX-License-Identifier: MIT
/**
 * Processor module that transforms data in the pipeline.
 *
 * @contract
 *   role: processing
 *   pattern: chain_of_responsibility
 */

import { BaseModule } from './imodule';

type TransformFn = (data: unknown) => unknown;

/**
 * Transforms data using a configurable function.
 *
 * @contract
 *   thread_safety: safe_after_start
 *   invariants:
 *     - transform function must be pure
 */
export class Processor extends BaseModule {
    private readonly transform: TransformFn;
    private processedCount: number = 0;

    constructor(transform?: TransformFn) {
        super();
        this.transform = transform || ((x: unknown) => (x as number) * 2);
    }

    /**
     * Transform data and forward to next module.
     *
     * @contract
     *   precondition: data is not null
     *   postcondition: result == transform(data)
     */
    process(data: unknown): unknown {
        const result = this.transform(data);
        this.processedCount++;
        return this.forward(result);
    }

    getProcessedCount(): number {
        return this.processedCount;
    }
}

File: test_projects/python_pipeline/processor.py
# SPDX-License-Identifier: MIT
"""
Processor module that transforms data in the pipeline.

@contract:
  role: processing
  pattern: chain_of_responsibility
"""

from typing import Any, Callable, Optional

from imodule import IModule


class Processor(IModule):
    """
    Transforms data using a configurable function.

    @contract:
      thread_safety: safe_after_start
      invariants:
        - transform function must be pure
    """

    def __init__(self, transform: Optional[Callable[[Any], Any]] = None) -> None:
        super().__init__()
        self._transform = transform or (lambda x: x * 2)
        self._processed_count = 0

    def process(self, data: Any) -> Any:
        """
        Transform data and forward to next module.

        @contract:
          precondition: data is not None
          postcondition: result == transform(data)
        """
        result = self._transform(data)
        self._processed_count += 1
        return self._forward(result)

    @property
    def processed_count(self) -> int:
        """Return the number of items processed."""
        return self._processed_count
